### PR TITLE
apiserver : Ensure required parameters are not empty

### PR DESCRIPTION
**Pull request by davidjumani**
### Description

Addon to https://github.com/apache/cloudstack/pull/5135
Ensures that any required parameter is not empty / null

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### How Has This Been Tested?

```
(localhost) 🐱 > create domain 
💩 Missing required parameters:  name
(localhost) 🐱 > create domain name= 
🙈 Error: (HTTP 431, error code 4350) Empty or null value provided for API arg: name
```

### CODE REVIEW

**Comment by davidjumani** [comment]
@blueorangutan package

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by davidjumani** [comment]
@blueorangutan test

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan test

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan test matrix

**Comment by davidjumani** [comment]
@blueorangutan test centos7 vmware-65u2

File: api/src/main/java/org/apache/cloudstack/api/command/admin/network/CreateNetworkOfferingCmd.java [hunk 1]
```diff
@@ -94,7 +94,6 @@
 
     @Parameter(name = ApiConstants.SUPPORTED_SERVICES,
             type = CommandType.LIST,
-            required = true,
```
Comment (GabrielBrascher) [inline api/src/main/java/org/apache/cloudstack/api/command/admin/network/CreateNetworkOfferingCmd.java:97]: Is there any update on the API command [createNetworkOffering](https://cloudstack.apache.org/api/apidocs-4.15/apis/createNetworkOffering.html) that I am not aware of?
As far as I know, `supportedservices` is required.

**Review by GabrielBrascher — COMMENTED** [review]
Thanks for the PR @davidjumani.
Overall code LGTM; I'm just curious about the removal of the `required = true` from `supportedservices`.

File: api/src/main/java/org/apache/cloudstack/api/command/admin/network/CreateNetworkOfferingCmd.java (hunk 1 above)
Comment (davidjumani) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/network/CreateNetworkOfferingCmd.java:97]: For L2 network offerings, there are cases when there are no services, and an empty list is sent, so marking it as not required

Comment (GabrielBrascher) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/network/CreateNetworkOfferingCmd.java:97]: Aha, looks like a legacy restriction that no longer is necessary.
@davidjumani thanks for the clarification  :+1:

**Review by GabrielBrascher — APPROVED** [review]
Code LGTM

File: server/src/main/java/com/cloud/api/dispatch/ParamProcessWorker.java [hunk 2]
```diff
@@ -159,6 +159,9 @@ public void processParameters(final BaseCmd cmd, final Map params) {
                 }
                 continue;
             }
+            if (parameterAnnotation.required()){
+                validateNonEmptyString(paramObj, parameterAnnotation.name());
```
Comment (rhtyd) [inline server/src/main/java/com/cloud/api/dispatch/ParamProcessWorker.java:163]: LGTM - my only worry is if it causes a regression as it create a new behaviour (nit)

**Review by shwstppr — APPROVED** [review]
LGTM. Tested locally with simulator,

For required empty params:
```
(local) 🐱 > create domain name=
🙈 Error: (HTTP 431, error code 4350) Empty or null value provided for API arg: name
```
For non-required empty params:
```
(local) 🐱 > create volume name= zoneid=1b1aecd6-4ee1-4a05-a02c-e34625387a1a diskofferingid=01c41386-2101-48eb-b68e-610d1b5252d2 
{
  "volume": {
    "account": "admin",
    "created": "2021-06-29T14:41:16+0530",
    "destroyed": false,
    "diskofferingdisplaytext": "Small Disk, 5 GB",
    "diskofferingid": "01c41386-2101-48eb-b68e-610d1b5252d2",
    "diskofferingname": "Small",
    "displayvolume": true,
    "domain": "ROOT",
    "domainid": "1f376e9f-d8b4-11eb-bd2c-645d8651f45a",
    "hypervisor": "None",
    "id": "0c4a7918-d2eb-42d3-ada9-b2880402a36c",
    "isextractable": true,
    "jobid": "ca5a09cf-10d8-49ac-ad00-cb01bf57750c",
    "jobstatus": 0,
    "name": "8f16fe2e-cd2d-4705-b4fc-f138b726edbd",
    "provisioningtype": "thin",
    "quiescevm": false,
    "size": 5368709120,
    "state": "Allocated",
    "storagetype": "shared",
    "supportsstoragesnapshot": false,
    "tags": [],
    "type": "DATADISK",
    "zoneid": "1b1aecd6-4ee1-4a05-a02c-e34625387a1a",
    "zonename": "Sandbox-simulator"
  }
}
```

**Comment by DaanHoogland** [comment]
rekicked the final failing travis job (5) but i think we can merge

**Comment by sureshanaparti** [comment]
@davidjumani I understand that optional parameters passed in the API are ignored, but these are still not null. In any case, the parameter value return based on not null check, there can be issue (for example, not null Boolean params can be interpreted as true/false (whatever default value for Boolean in java), and the default parameter value is the opposite).

**Comment by davidjumani** [comment]
@sureshanaparti There are no changes to optional parameters, only parameters with required=true will now have to not be empty

**Comment by sureshanaparti** [comment]
> @sureshanaparti There are no changes to optional parameters, only parameters with required=true will now have to not be empty

As the required parameters are validated here, before validating them through the _validations_ annotation, this annotation `validations = {ApiArgValidator.NotNullOrEmpty}` is no longer valid for required parameters.

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan test

**Pull request merged**